ompi/tools/mpirun/main.c

```diff
@@ -14,100 +14,118 @@
-        if (0 == strcmp(argv[m], "--prefix")) {
-            opal_asprintf(&pfx, "%s%s", argv[m+1], "/bin");
-        }
-    }
-
-    if (NULL != pfx) {
-        /* "Parse" the param, aka remove superfluous path_sep. */
-        param_len = strlen(pfx);
-        while (0 == strcmp(OPAL_PATH_SEP, &(pfx[param_len - 1]))) {
-            pfx[param_len - 1] = '\0';
-            param_len--;
-            if (0 == param_len) {
-                fprintf(stderr, "A prefix was supplied to mpirun that only contained slashes.\n"
-                        "This is a fatal error; mpirun will now abort.\nNo processes were launched.\n");
-                exit(1);
-            }
-        }
-    } else if (opal_path_is_absolute(argv[0])) {
-        /* Check if called with fully-qualified path to mpirun.
-         * (Note: Put this second so can override with --prefix (above). */
-        pfx = opal_dirname(argv[0]);
+    /* as a special case, if OPAL_PREFIX was set and either PRRTE or
+     * PMIx are internal builds, set their prefix variables as well */
+    if (NULL != opal_prefix) {
 #if OMPI_USING_INTERNAL_PRRTE
-    } else {
-        /* in case --enable-prefix-by-default was given */
-        mca_base_framework_open(&opal_installdirs_base_framework, 0);  // fill in the installdirs
-        if (NULL != opal_install_dirs.bindir) {
-            pfx = strdup(opal_install_dirs.bindir);
-        }
+        setenv("PRTE_PREFIX", opal_prefix, 1);
 #endif
-    }
-
-    if (NULL == pfx) {
-        truepath = opal_path_findv("prterun", X_OK, environ, NULL);
-#if !OMPI_USING_INTERNAL_PRRTE
-        // if OMPI_PRTERUN_PATH is available, try that
-        // for external builds if the user didn't explictly
-        // add a prefix and it isn't in the users path.
-        if((NULL == truepath) && (0 != strlen(OMPI_PRTERUN_PATH))) {
-            truepath = opal_os_path(0, OMPI_PRTERUN_PATH, "prterun", NULL);
-        }
+#if OPAL_USING_INTERNAL_PMIX
+        setenv("PMIX_PREFIX", opal_prefix, 1);
 #endif
-    } else {
-        truepath = opal_os_path(0, pfx, "prterun", NULL);
-        free(pfx);
     }
 
-    if (NULL == truepath) {
-        fprintf(stderr, "prterun executable could not be found - unable to run\n");
+    full_prterun_path = find_prterun();
+    if (NULL == full_prterun_path) {
+        opal_show_help("help-mpirun.txt", "no-prterun-found", 1);
         exit(1);
     }
 
-    execve(truepath, pargs, environ);
-    fprintf(stderr, "The mpirun (\"%s\") cmd failed to exec its actual executable - your application will NOT execute. Error: %s\n",
-                     truepath ? truepath : "NULL", strerror(errno));
+    /* calling mpirun (and now prterun) with a full path has a special
+     * meaning in terms of -prefix behavior, so copy that behavior
+     * into prterun */
+    if (opal_path_is_absolute(argv[0])) {
+        opal_argv_append_nosize(&prterun_args, full_prterun_path);
+    } else {
+        opal_argv_append_nosize(&prterun_args, "prterun");
+    }
```

Comment:
I'm a little confused by this.  For example, if the user set the env var `$OMPI_PRTERUN`, but then didn't specify an absolute path to `mpirun`, it looks like we still just exec `prterun`, not the value of `$OMPI_PRTERUN`.  Is that what was intended?

Reply:
no, that's not what we do.  Execv takes two arguments; path to the executable to execute and an argv array.  The argv array includes what will become argv[0] in the child.  The full path and the argv[0] do not have to match.

Unfortunately, mpirun's argv[0] is parsed to figure out whether to have --prefix behavior.  So if the user specified mpirun as a relative path, we do the same for argv[0] so that we cause the same prefix behavior, but we still pass the full path for the executable to execute.

Reply:
Got it: I missed the 2-arg execv.